instrumentation/graphql-java/graphql-java-20.0/library/src/main/java/io/opentelemetry/instrumentation/graphql/v20_0/OpenTelemetryInstrumentation.java

```diff
@@ -34,23 +77,150 @@ public InstrumentationState createState(InstrumentationCreateStateParameters par
   @Override
   public InstrumentationContext<ExecutionResult> beginExecution(
       InstrumentationExecutionParameters parameters, InstrumentationState rawState) {
+
     OpenTelemetryInstrumentationState state = ofState(rawState);
-    return helper.beginExecution(state);
+    Context parentContext = Context.current();
+
+    if (!executionInstrumenter.shouldStart(parentContext, state)) {
+      return SimpleInstrumentationContext.noOp();
+    }
+
+    Context context = executionInstrumenter.start(parentContext, state);
+    state.setContext(context);
+
+    return SimpleInstrumentationContext.whenCompleted(
+        (result, throwable) -> {
+          Span span = Span.fromContext(context);
+
+          for (GraphQLError error : result.getErrors()) {
+            AttributesBuilder attributes = Attributes.builder();
+            attributes.put(
+                ExceptionAttributes.EXCEPTION_TYPE, String.valueOf(error.getErrorType()));
+            attributes.put(ExceptionAttributes.EXCEPTION_MESSAGE, error.getMessage());
+
+            span.addEvent("exception", attributes.build());
+          }
+
+          executionInstrumenter.end(context, state, result, throwable);
+        });
   }
 
   @Override
   public InstrumentationContext<ExecutionResult> beginExecuteOperation(
       InstrumentationExecuteOperationParameters parameters, InstrumentationState rawState) {
+
     OpenTelemetryInstrumentationState state = ofState(rawState);
-    return helper.beginExecuteOperation(parameters, state);
+    Span span = Span.fromContext(state.getContext());
+
+    OperationDefinition operationDefinition =
+        parameters.getExecutionContext().getOperationDefinition();
+    OperationDefinition.Operation operation = operationDefinition.getOperation();
+
+    String operationType = operation.name().toLowerCase(Locale.ROOT);
+    String operationName = operationDefinition.getName();
+
+    String spanName = operationType;
+    if (operationName != null && !operationName.isEmpty()) {
+      spanName += " " + operationName;
+    }
+    span.updateName(spanName);
+
+    state.setOperation(operation);
+    state.setOperationName(operationName);
+
+    Node<?> node = operationDefinition;
+    if (sanitizeQuery) {
+      node = sanitize(node);
+    }
+    state.setQuery(AstPrinter.printAst(node));
```

Comment:
I'm still yet to verify this but suspect this block is the source of performance regressions in our GraphQL endpoint in our production environment. I'm unable to turn this off as this needs to be sanitised.

We are seeing regressions of around 10-100+ milliseconds added to our GraphQL operations when the OpenTelemetry GraphQL instrumentation is turned on. I suspect traversing the AST is causing the regression with larger GraphQL documents resulting in larger regressions.

I think this block needs to be surrounded with a check on `span.isRecording()`.

Reply:
Alternatively we could find a way to off load this onto another thread to do this async as to not block the execution of the GraphQL operation.

Reply:
If you can then please do verify. If we can reproduce this then we can figure out a better strategy to handle it. If the query is really large then perhaps we can only handle the start of the query and truncate most of it.